File: client/node/src/bucket.ts
import { StorageClient } from "./gen/proto/storage/v2/storage_grpc_pb";
import {
  StorageReadRequest,
  StorageWriteRequest,
  StorageDeleteRequest,
  StorageListBlobsRequest,
  StorageExistsRequest,
  StoragePreSignUrlRequest,
} from "./gen/proto/storage/v2/storage_pb";
import { Duration } from "google-protobuf/google/protobuf/duration_pb";

export enum Mode {
  Read = "READ",
  Write = "WRITE",
}

export interface PresignUrlOptions {
  mode: Mode;
  expiry: number; // Duration in seconds
}

export class Bucket {
  private name: string;
  private storageClient: StorageClient;

  constructor(storageClient: StorageClient, bucketName: string) {
    this.name = bucketName;
    this.storageClient = storageClient;
  }

  /**
   * Read a file from the bucket
   */
  async read(key: string): Promise<Buffer> {
    const request = new StorageReadRequest();
    request.setBucketName(this.name);
    request.setKey(key);

    return new Promise((resolve, reject) => {
      this.storageClient.read(request, (error, response) => {
        if (error) {
          reject(
            new Error(
              `Failed to read file from the ${this.name} bucket: ${error.message}`
            )
          );
          return;
        }
        if (!response) {
          reject(
            new Error(
              `Failed to read file from the ${this.name} bucket: No response received`
            )
          );
          return;
        }
        resolve(Buffer.from(response.getBody_asU8()));
      });
    });
  }

  /**
   * Write a file to the bucket
   */
  async write(key: string, data: Buffer): Promise<void> {
    const request = new StorageWriteRequest();
    request.setBucketName(this.name);
    request.setKey(key);
    request.setBody(data);

    return new Promise((resolve, reject) => {
      this.storageClient.write(request, (error) => {
        if (error) {
          reject(new Error(`Failed to write file to bucket: ${error.message}`));
          return;
        }
        resolve();
      });
    });
  }

  /**
   * Delete a file from the bucket
   */
  async delete(key: string): Promise<void> {
    const request = new StorageDeleteRequest();
    request.setBucketName(this.name);
    request.setKey(key);

    return new Promise((resolve, reject) => {
      this.storageClient.delete(request, (error) => {
        if (error) {
          reject(
            new Error(`Failed to delete file from bucket: ${error.message}`)
          );
          return;
        }
        resolve();
      });
    });
  }

  /**
   * List files in the bucket with a given prefix
   */
  async list(prefix: string): Promise<string[]> {
    const request = new StorageListBlobsRequest();
    request.setBucketName(this.name);
    request.setPrefix(prefix);

    return new Promise((resolve, reject) => {
      this.storageClient.listBlobs(request, (error, response) => {
        if (error) {
          reject(new Error(`Failed to list files in bucket: ${error.message}`));
          return;
        }
        if (!response) {
          reject(
            new Error(`Failed to list files in bucket: No response received`)
          );
          return;
        }
        const blobs = response.getBlobsList();
        resolve(blobs.map((blob) => blob.getKey()));
      });
    });
  }

  /**
   * Check if a file exists in the bucket
   */
  async exists(key: string): Promise<boolean> {
    const request = new StorageExistsRequest();
    request.setBucketName(this.name);
    request.setKey(key);

    return new Promise((resolve, reject) => {
      this.storageClient.exists(request, (error, response) => {
        if (error) {
          reject(
            new Error(
              `Failed to check if file exists in bucket: ${error.message}`
            )
          );
          return;
        }
        if (!response) {
          reject(
            new Error(
              `Failed to check if file exists in bucket: No response received`
            )
          );
          return;
        }
        resolve(response.getExists());
      });
    });
  }

  /**
   * Get a presigned URL for downloading a file from the bucket
   */
  async getDownloadUrl(
    key: string,
    options?: Partial<PresignUrlOptions>
  ): Promise<string> {
    const defaultOptions: PresignUrlOptions = {
      mode: Mode.Read,
      expiry: 300, // 5 minutes in seconds
    };

    const opts = { ...defaultOptions, ...options };
    return this.preSignUrl(key, opts);
  }

  /**
   * Get a presigned URL for uploading a file to the bucket
   */
  async getUploadUrl(
    key: string,
    options?: Partial<PresignUrlOptions>
  ): Promise<string> {
    const defaultOptions: PresignUrlOptions = {
      mode: Mode.Write,
      expiry: 300, // 5 minutes in seconds
    };

    const opts = { ...defaultOptions, ...options };
    return this.preSignUrl(key, opts);
  }

  private async preSignUrl(
    key: string,
    options: PresignUrlOptions
  ): Promise<string> {
    const request = new StoragePreSignUrlRequest();
    request.setBucketName(this.name);
    request.setKey(key);
    request.setOperation(
      options.mode === Mode.Read
        ? StoragePreSignUrlRequest.Operation.READ
        : StoragePreSignUrlRequest.Operation.WRITE
    );

    const duration = new Duration();
    duration.setSeconds(options.expiry);
    request.setExpiry(duration);

    return new Promise((resolve, reject) => {
      this.storageClient.preSignUrl(request, (error, response) => {
        if (error) {
          reject(
            new Error(`Failed to get presigned URL for file: ${error.message}`)
          );
          return;
        }
        if (!response) {
          reject(
            new Error(
              `Failed to get presigned URL for file: No response received`
            )
          );
          return;
        }
        resolve(response.getUrl());
      });
    });
  }
}

File: client/go/suga/bucket.go
package suga

import (
	"context"
	"fmt"
	"time"

	storagepb "github.com/nitrictech/nitric/proto/storage/v2"
	"google.golang.org/protobuf/types/known/durationpb"
)

// Bucket provides methods for interacting with cloud storage buckets
type Bucket struct {
	name          string
	storageClient storagepb.StorageClient
}

// Read a file from the bucket
func (c *Bucket) Read(key string) ([]byte, error) {
	ctx := context.Background()

	req := &storagepb.StorageReadRequest{
		BucketName: c.name,
		Key:        key,
	}

	res, err := c.storageClient.Read(ctx, req)
	if err != nil {
		return nil, fmt.Errorf("failed to read file from the %s bucket: %w", c.name, err)
	}

	return res.Body, nil
}

// Write a file to the bucket
func (c *Bucket) Write(key string, data []byte) error {
	ctx := context.Background()

	req := &storagepb.StorageWriteRequest{
		BucketName: c.name,
		Key:        key,
		Body:       data,
	}

	_, err := c.storageClient.Write(ctx, req)
	if err != nil {
		return fmt.Errorf("failed to write file to bucket: %w", err)
	}

	return nil
}

// Delete a file from the bucket
func (c *Bucket) Delete(key string) error {
	ctx := context.Background()

	req := &storagepb.StorageDeleteRequest{
		BucketName: c.name,
		Key:        key,
	}

	_, err := c.storageClient.Delete(ctx, req)
	if err != nil {
		return fmt.Errorf("failed to delete file from bucket: %w", err)
	}

	return nil
}

// List files in the bucket with a given prefix
func (c *Bucket) List(prefix string) ([]string, error) {
	ctx := context.Background()

	req := &storagepb.StorageListBlobsRequest{
		BucketName: c.name,
		Prefix:     prefix,
	}

	res, err := c.storageClient.ListBlobs(ctx, req)
	if err != nil {
		return nil, fmt.Errorf("failed to list files in bucket: %w", err)
	}

	keys := make([]string, len(res.Blobs))
	for i, blob := range res.Blobs {
		keys[i] = blob.Key
	}

	return keys, nil
}

// Exists checks if a file exists in the bucket
func (c *Bucket) Exists(key string) (bool, error) {
	ctx := context.Background()

	req := &storagepb.StorageExistsRequest{
		BucketName: c.name,
		Key:        key,
	}

	res, err := c.storageClient.Exists(ctx, req)
	if err != nil {
		return false, fmt.Errorf("failed to check if file exists in bucket: %w", err)
	}

	return res.Exists, nil
}

type Mode int

const (
	ModeRead Mode = iota
	ModeWrite
)

type presignUrlOptions struct {
	mode   Mode
	expiry time.Duration
}

type PresignUrlOption func(opts *presignUrlOptions)

// WithPresignUrlExpiry sets the expiry duration for presigned URLs
func WithPresignUrlExpiry(expiry time.Duration) PresignUrlOption {
	return func(opts *presignUrlOptions) {
		opts.expiry = expiry
	}
}

func getPresignUrlOpts(mode Mode, opts ...PresignUrlOption) *presignUrlOptions {
	defaultOpts := &presignUrlOptions{
		mode:   mode,
		expiry: time.Minute * 5,
	}

	for _, opt := range opts {
		opt(defaultOpts)
	}

	return defaultOpts
}

func (c *Bucket) preSignUrl(key string, opts *presignUrlOptions) (string, error) {
	ctx := context.Background()

	op := storagepb.StoragePreSignUrlRequest_READ

	if opts.mode == ModeWrite {
		op = storagepb.StoragePreSignUrlRequest_WRITE
	}

	req := &storagepb.StoragePreSignUrlRequest{
		BucketName: c.name,
		Key:        key,
		Operation:  op,
		Expiry:     durationpb.New(opts.expiry),
	}

	res, err := c.storageClient.PreSignUrl(ctx, req)
	if err != nil {
		return "", fmt.Errorf("failed to get presigned URL for file: %w", err)
	}

	return res.Url, nil
}

// GetDownloadURL returns a presigned URL for downloading a file from the bucket, with a limited expiry time
func (c *Bucket) GetDownloadURL(key string, opts ...PresignUrlOption) (string, error) {
	optsWithDefaults := getPresignUrlOpts(ModeRead, opts...)

	return c.preSignUrl(key, optsWithDefaults)
}

// GetUploadURL returns a presigned URL for uploading a file to the bucket, with a limited expiry time
func (c *Bucket) GetUploadURL(key string, opts ...PresignUrlOption) (string, error) {
	optsWithDefaults := getPresignUrlOpts(ModeWrite, opts...)

	return c.preSignUrl(key, optsWithDefaults)
}

// NewBucket creates a new client interactive with a named bucket
func NewBucket(storageClient storagepb.StorageClient, bucketName string) *Bucket {
	return &Bucket{
		storageClient: storageClient,
		name:          bucketName,
	}
}
